### PR TITLE
Feed creation refactoring

### PR DESCRIPTION
**Pull request by splitbrain**
This breaks up the humongous functions from feed.php into multiple classes. To keep compatibility with existing Plugin events, the basic principle of how the feed is assembled has not been changed:

* depending on the given mode and other options lose arrays of items are gathered
* these items are then converted (again based on the various options) into proper FeedItems
* the FeedItems are then added to the Feed

The conversion from loosely typed item data to something more structured is now done by the FeedItemProcessor classes. Some very basic tests have been added. It does not cover erverything but covers more than before (which was nothing).

Manual testing before merging this is highly recommended. I am not confident that I ported over everything correctly.

No new features have been added, but especially media support could and should be improved in the future.

This should address:

* #4030 - undefined keys should be handled by the processors now
* #4031 - this PR replaces it
* #3970 - undefined keys should be handled by the processors now
* #3989 - This PR replaces it
* #3953 - abstracts should be returned correctly, but the issue seems somewhat bogus anyway
* #2239 - tagline is set as subtitle
* #2960 - date is passed as timestamp now, upstream update will be done in master later

### CODE REVIEW

**Comment by soatok** [comment]
GitHub Actions runs on Azure, IIRC, and your tests ma be assuming some guarantees about file modification times that aren't true in the cloud environment they're executed within.

Out of curiosity, does `clearstatcache()` affect this weirdness?

**Comment by splitbrain** [comment]
> your tests may be assuming some guarantees about file modification times that aren't true in the cloud environment 

We occasionally have off-by-one-second errors that come and go, but here it seems to be too consistent and too large of a difference to be a fluke.

> Out of curiosity, does clearstatcache() affect this weirdness?

That was my first idea I added a bunch of clearstatcaches() in my first force-push. No change.

**Comment by soatok** [comment]
I made a draft PR (#4160) to test a hypothesis. Please approve it to run when you have a chance.

**Comment by soatok** [comment]
Okay, I think this failure is actually because filemtime is a parameter on the test parameter.

To test my hypothesis, I [added some `var_dump()` calls](https://github.com/soatok/dokuwiki/actions/runs/7666830054/job/20895411707#step:7:11) to the data provider to get the exact timestamps (down to the microsecond) of each `yield`.

Before the first `yiel;d`:

```
array(1) {
  ["now"]=>
  float(1706264658.595)
}
```

Before the second yield:

```
array(2) {
  ["early stat"]=>
  array(26) {
/* snip */
    ["mtime"]=>
    int(1706264658)
    ["ctime"]=>
    int(1706264658)
/* snip */
  ["now"]=>
  float(1706264658.5951)
}
```

And before the third yield:

```
array(2) {
  ["early stat"]=>
  array(26) {
/* snip */
    ["mtime"]=>
    int(1706264658)
    ["ctime"]=>
    int(1706264658)
/* snip */
  ["now"]=>
  float(1706264658.5952)
}
```

Okay, we're almost there. I still get an error:

> Failed asserting that 1706264737 matches expected 1706264658.

However, this appears to be **the total elapsed time** for PHPUnit from initialization to when this specific test is run. That is to say, the var_dump() calls in the generator are all invoked before the actual test suite begins running.

A subsequent test that adds a microtime() dump at the top of the test code confirms this hypothesis: https://github.com/soatok/dokuwiki/actions/runs/7667005050/job/20895926731

## TL;DR - The generator is exhausted as PHPUnit warms up, not on-demand. 

Fix: Change the filemtime from a PHPUnit data provider to a simple private method that contains the necessary `yield`s and you should see it behave as you desire.

**Comment by soatok** [comment]
#4161 contains a proposed fix.

**Comment by splitbrain** [comment]
:tada:  Awesome work @soatok! Thanks for helping out.
